cli/compose/types/types.go

```diff
@@ -70,14 +141,69 @@ func (cd ConfigDetails) LookupEnv(key string) (string, bool) {
 
 // Config is a full compose file configuration
 type Config struct {
-	Filename string `yaml:"-"`
-	Version  string
-	Services Services
-	Networks map[string]NetworkConfig   `yaml:",omitempty"`
-	Volumes  map[string]VolumeConfig    `yaml:",omitempty"`
-	Secrets  map[string]SecretConfig    `yaml:",omitempty"`
-	Configs  map[string]ConfigObjConfig `yaml:",omitempty"`
-	Extras   map[string]interface{}     `yaml:",inline"`
+	Filename string                     `yaml:"-" json:"-"`
+	Version  string                     `json:"version"`
+	Services Services                   `json:"services"`
+	Networks map[string]NetworkConfig   `yaml:",omitempty" json:"networks,omitempty"`
+	Volumes  map[string]VolumeConfig    `yaml:",omitempty" json:"volumes,omitempty"`
+	Secrets  map[string]SecretConfig    `yaml:",omitempty" json:"secrets,omitempty"`
+	Configs  map[string]ConfigObjConfig `yaml:",omitempty" json:"configs,omitempty"`
+	Extras   map[string]interface{}     `yaml:",inline", json:"-"`
+}
+
+// MarshalJSON makes Config implement json.Marshaler
+func (c Config) MarshalJSON() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	fmt.Fprintf(buf, `{
+  "version": %q,
+  "services":`, c.Version)
+	js, err := json.MarshalIndent(c.Services, "  ", "  ")
+	if err != nil {
+		return nil, err
+	}
+	buf.Write(js)
+	if len(c.Networks) > 0 {
+		buf.WriteString(",\n  \"networks\": ")
```

Comment:
I don't think we should do the `\n` or even that "string" manually. We should be able to build a struct or map that `encoding/json` can marshal in JSON (and thus the way it wants to be…). I feel this part of the code is way more complicated than it should…

Reply:
The following code should do the trick nicely

```go

func (c Config) MarshalJSON() ([]byte, error) {
	m := map[string]interface{}{
		"version":  c.Version,
		"services": c.Services,
	}
	if len(c.Networks) > 0 {
		m["networks"] = c.Networks
	}
	if len(c.Volumes) > 0 {
		m["volumes"] = c.Volumes
	}
	if len(c.Secrets) > 0 {
		m["secrets"] = c.Secrets
	}
	if len(c.Configs) > 0 {
		m["configs"] = c.Configs
	}
	for k, v := range c.Extras {
		m[k] = v
	}
	return json.Marshal(m)
}
```

Reply:
I tried that, but [maps get shuffled](https://nathanleclaire.com/blog/2014/04/27/a-surprising-feature-of-golang-that-colored-me-impressed/) when marshaled to JSON, so your solution doesn't preserve the order of the fields, which is an important feature of the `struct` type.

Reply:
I don't think we should care about the order, it will be marshaled correctly anyway… the `json` format is just data…

Reply:
It's data that is likely to be read and modified by humans, who do care about semantic order.

Reply:
What about the YAML marshaller then? I think that the YAML has more chances to be read by humans (because of the compose files) than the JSON version.

Reply:
@silvin-lubecki What about it? The YAML output is always ordered the same way based on the ordering of the struct.

Reply:
JSON is very rarely used to author Compose files as far as I'm aware: GitHub shows [237](https://github.com/search?q=in%3Apath+docker-compose.json+extension%3Ajson&type=Code) vs [~300,000](https://github.com/search?q=in%3Apath+docker-compose.yml+extension%3Ayml&type=Code) for YAML.

I think given that it's reasonable to assume where JSON is used it's more of a machine to machine format, and ordering isn't overly important as long as it's deterministic. The usecase I've seen (and done myself) here is JSON is increasingly available in the standard library of languages, YAML (because it's crazy) isn't.

Reply:
Thanks @garethr ! Updated accordingly.